### PR TITLE
Feat: Enchant System

### PR DESCRIPTION
**Pull request by Zintixx**
- Implements enchant system
- Currently does not use the correct values for item/ingredient costs.

### CODE REVIEW

File: MapleServer2/PacketHandlers/Game/ItemEnchantHandler.cs [hunk 1]
```diff
@@ -43,54 +62,273 @@ public override void Handle(GameSession session, PacketReader packet)
...
+        EnchantType type = (EnchantType) packet.ReadByte();
         long itemUid = packet.ReadLong();
 
         Item item = session.Player.Inventory.GetByUid(itemUid);
+        if (item is null || item.DisableEnchant || item.EnchantLevel >= 15)
+        {
+            session.Send(ItemEnchantPacket.Notice((short) ItemEnchantError.ItemCannotBeEnchanted));
+            return;
+        }
+        if (item.Type is ItemType.Necklace or ItemType.Belt or ItemType.Earring or ItemType.Ring or ItemType.Shield or ItemType.Spellbook)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/ItemEnchantHandler.cs:74]: Empty line between 73 and 74

**Review by AlanMorel — APPROVED** [review]

File: MapleServer2/PacketHandlers/Game/ItemEnchantHandler.cs [hunk 2]
```diff
@@ -43,54 +62,273 @@ public override void Handle(GameSession session, PacketReader packet)
...
+            case 10:
+                return 3;
+            case 11:
+            case 12:
+                return 5;
+            case 13:
+            case 14:
+                return 8;
+        }
+        return 0;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/ItemEnchantHandler.cs:332]:
```suggestion
        return currentEnchantLevel switch
        {
            0 or 1 or 2 => 1,
            3 or 4 or 5 => 2,
            6 or 7 or 8 => 4,
            9 => 5,
            10 => 3,
            11 or 12 => 5,
            13 or 14 => 8,
            _ => 0
        };
```

File: MapleServer2/Packets/ItemEnchantPacket.cs [hunk 3]
```diff
@@ -1,97 +1,132 @@
...
-            pWriter.WriteShort();
-            pWriter.WriteInt();
-            pWriter.Write(0f);
+            pWriter.Write(stats.Attribute);
+            pWriter.WriteInt(stats.AddValue);
+            pWriter.WriteFloat(stats.AddRate);
         }
 
-        if (type == 1)
+        if (type == EnchantType.Ophelia)
```
Comment (AngeloTadeucci) [inline MapleServer2/Packets/ItemEnchantPacket.cs:44]:
```suggestion
        if (type is EnchantType.Ophelia)
```

**Review by AngeloTadeucci — APPROVED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Review by darkvergus — APPROVED** [review]

File: MapleServer2/PacketHandlers/Game/ItemEnchantHandler.cs [hunk 4]
```diff
@@ -43,54 +62,273 @@ public override void Handle(GameSession session, PacketReader packet)
 
     private static void HandleBeginEnchant(GameSession session, PacketReader packet)
     {
-        byte type = packet.ReadByte();
+        EnchantType type = (EnchantType) packet.ReadByte();
         long itemUid = packet.ReadLong();
 
         Item item = session.Player.Inventory.GetByUid(itemUid);
+        if (item is null || item.DisableEnchant || item.EnchantLevel >= 15)
```
[the remark this answers is not in the document]
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/ItemEnchantHandler.cs:69]: Enchant != Limit Break

**Pull request merged**